File: 2948-make-lexicographically-smallest-array-by-swapping-elements/2948-make-lexicographically-smallest-array-by-swapping-elements.js
/**
 * @param {number[]} nums
 * @param {number} limit
 * @return {number[]}
 */
var lexicographicallySmallestArray = function(nums, limit) {

    const n = nums.length;

    // [value, originalIndex]
    const arr = [];

    for (let i = 0; i < n; i++) {
        arr.push([nums[i], i]);
    }

    // Sort by value
    arr.sort((a, b) => a[0] - b[0]);

    let start = 0;

    while (start < n) {

        let end = start;

        // Find group
        while (
            end + 1 < n &&
            arr[end + 1][0] - arr[end][0] <= limit
        ) {
            end++;
        }

        // Get original indices
        const indices = [];

        for (let i = start; i <= end; i++) {
            indices.push(arr[i][1]);
        }

        // Sort indices
        indices.sort((a, b) => a - b);

        // Assign smallest values to smallest indices
        for (let i = start; i <= end; i++) {
            nums[indices[i - start]] = arr[i][0];
        }

        start = end + 1;
    }

    return nums;
};

File: 2948-make-lexicographically-smallest-array-by-swapping-elements/2948-make-lexicographically-smallest-array-by-swapping-elements.py
class Solution:
    def lexicographicallySmallestArray(self, nums, limit):

        n = len(nums)

        # (value, original index)
        arr = [(nums[i], i) for i in range(n)]

        # Sort by value
        arr.sort()

        start = 0

        while start < n:

            end = start

            # Find current group
            while (
                end + 1 < n
                and arr[end + 1][0] - arr[end][0] <= limit
            ):
                end += 1

            # Original indices
            indices = [
                arr[i][1]
                for i in range(start, end + 1)
            ]

            # Sort indices
            indices.sort()

            # Assign smallest values to smallest indices
            for i in range(start, end + 1):
                nums[indices[i - start]] = arr[i][0]

            start = end + 1

        return nums